pkg/config/config.go

```diff
@@ -1588,60 +1690,100 @@ func (c *Config) PublicFields() (map[string]string, error) {
 	return result, nil
 }
 
-func LoadConfig(configFile string) (cfg *Config, err error) {
+func LoadConfig(configFile string) (*Config, error) {
 	var filesToCheck []string
 	if configFile != "" {
 		filesToCheck = append(filesToCheck, configFile)
 	}
-	filesToCheck = append(filesToCheck, defaultConfigFiles...)
 
-	cfg = NewConfig()
+	filesToCheck = append(filesToCheck, defaultConfigFiles...)
+	cfg := NewConfig()
 	cfgMetadata, err := config_loader.LoadYamlConfig(cfg, filesToCheck...)
 	if err != nil {
-		err = fmt.Errorf("unable to parse configuration file %s: %s", configFile, err)
-		return
+		err = fmt.Errorf("%w, %s: %s", ErrUnableToParseConfigFile, configFile, err.Error())
+		return cfg, err
 	}
 
-	// After the config file has loaded,  override via any environment variables
-	configOverride(cfg)
+	if !cfg.Databind.IsEmpty() {
+		databindSources, errD := cfg.Databind.DataSources()
+		if errD != nil {
+			//nolint:wrapcheck
+			return cfg, fmt.Errorf("%w: %v", ErrDatabindApply, errD.Error())
+		}
+		templateConfig := NewConfig()
 
-	sources, err := cfg.Databind.DataSources()
-	if err != nil {
-		return
-	}
-	//_, err = databind.Fetch(sources)
-	vals, err := databind.Fetch(sources)
-	if err != nil {
-		return
-	}
-	if vals.VarsLen() > 0 {
-		cfg.Databind = databind.YAMLAgentConfig{}
-		matches, errD := databind.Replace(&vals, cfg)
+		_, errD = config_loader.LoadYamlConfig(templateConfig, filesToCheck...)
 		if errD != nil {
-			return
+			//nolint:wrapcheck
+			return cfg, fmt.Errorf("%w: %v", ErrDatabindApply, errD.Error())
 		}
 
-		if len(matches) != 1 {
-			err = fmt.Errorf("unexpected config file variables replacement amount")
-			return
+		dynamicConfig := DynamicConfig{
+			databindSources:        databindSources,
+			templateConfigMetadata: cfgMetadata,
+			templateConfig:         templateConfig,
+			refreshedConfig:        nil,
 		}
-		transformed := matches[0]
-		replacedCfg, ok := transformed.Variables.(*Config)
-		if !ok {
-			err = fmt.Errorf("unexpected config file variables replacement type")
-			return
+		cfg.dynamicConfig = &dynamicConfig
+
+		cfg, errD = applyDatabind(&dynamicConfig)
+		if errD != nil {
+			//nolint:wrapcheck
+			return cfg, fmt.Errorf("%w: %v", ErrDatabindApply, errD.Error())
 		}
-		cfg = replacedCfg
 	}
 
+	// After the config file has loaded, override via any environment variables
+	configOverride(cfg)
+
 	cfg.RunMode, cfg.AgentUser, cfg.ExecutablePath = runtimeValues()
 
 	// Move any other post processing steps that clean up or announce settings to be
 	// after both config file and env variable processing is complete. Need to review each of the items
 	// above and place each one at the bottom of this ordering
 	err = NormalizeConfig(cfg, *cfgMetadata)
 
-	return
+	return cfg, err
+}
+
+func applyDatabind(dynamicConfig *DynamicConfig) (*Config, error) {
+	var err error
+
+	if dynamicConfig == nil {
+		//nolint:wrapcheck
```

Comment:
still required?

Reply:
Yes, this is where it started to ask for those wraps for errors. Given that we add nolint, I restored the old shape of the code